mayastor/src/grpc/mayastor_grpc.rs

```diff
@@ -32,36 +33,75 @@ use crate::{
     nexus_uri::NexusBdevError,
     subsys::PoolConfig,
 };
+use futures::FutureExt;
 use nix::errno::Errno;
 use rpc::mayastor::*;
-use std::{convert::TryFrom, ops::Deref, time::Duration};
+use std::{convert::TryFrom, fmt::Debug, ops::Deref, time::Duration};
 use tonic::{Request, Response, Status};
 #[derive(Debug)]
 struct UnixStream(tokio::net::UnixStream);
 
+use ::function_name::named;
+use std::panic::AssertUnwindSafe;
+
+impl GrpcClientContext {
+    #[track_caller]
+    pub fn new<T>(req: &Request<T>, fid: &str) -> Self
+    where
+        T: Debug,
+    {
+        Self {
+            args: format!("{:?}", req.get_ref()),
+            id: fid.to_string(),
+        }
+    }
+}
+
 #[derive(Debug)]
 pub struct MayastorSvc {
     name: String,
     interval: Duration,
-    rw_lock: tokio::sync::RwLock<bool>,
+    rw_lock: tokio::sync::RwLock<Option<GrpcClientContext>>,
 }
 
 #[async_trait::async_trait]
-impl<F, R> Serializer<F, R> for MayastorSvc
+impl<F, T> Serializer<F, T> for MayastorSvc
 where
-    F: core::future::Future<Output = R> + Send + 'static,
-    R: Send,
+    T: Send + 'static,
+    F: core::future::Future<Output = Result<T, Status>> + Send + 'static,
 {
-    async fn locked(&self, f: F) -> R {
-        let mut lock = self.rw_lock.write().await;
-        trace!(?self.name, "locked");
-        assert!(!*lock);
-        *lock = true;
-        let output = f.await;
-        tokio::time::sleep(self.interval).await;
-        trace!(?self.name, "unlocked");
-        *lock = false;
-        output
+    async fn locked(&self, ctx: GrpcClientContext, f: F) -> Result<T, Status> {
+        let mut guard = self.rw_lock.write().await;
+
+        // Store context as a marker of to detect abnormal termination of the
+        // request. Even though AssertUnwindSafe() allows us to
+        // intercept asserts in underlying method strategies, such a
+        // situation can still happen when the high-level future that
+        // represents gRPC call at the highest level (i.e. the one created
+        // by gRPC server) gets cancelled (due to timeout or somehow else).
+        // This can't be properly intercepted by 'locked' function itself in the
+        // first place, so the state needs to be cleaned up properly
+        // upon subsequent gRPC calls.
+        if let Some(c) = guard.replace(ctx) {
+            warn!("{}: gRPC method timed out, args: {}", c.id, c.args);
```

Comment:
We did not timeout; the callee lost its patience, small nit but can be confusing for people when they read this? how about, "grpc client disappeared" it something like that?

Reply:
Sure thing, fixed.